### ISSUE TITLE
ConvGNP loss calculation

### ISSUE THREAD

**Issue by jonas-scholz123** [tom-andersson/deepsensor#10]
### Description

There is a bug in the calculation for the likelihood in convGNPs (i.e. using `likelihood="lowrank"`), that yields incorrect and very low losses.

For example, the following fit generated a loss of -3148, even though it is a very bad fit.
[loss_-3148.861328125.pdf](https://github.com/tom-andersson/deepsensor/files/11726814/loss_-3148.861328125.pdf)

This incorrect loss leads to large gradients being applied that inhibit the training process:
![gnp_issue](https://github.com/tom-andersson/deepsensor/assets/37850411/dab3b088-ea8a-4387-89d4-ab8e544db232)

### How to Reproduce

Train a convGNP using only one task to overfit to it. To do so efficiently, set the context as well as target sampling to `all`.

### Expected result:
The model perfectly overfits to the one task, with the mean matching target measurements exactly and the predicted standard deviation approaching 0.

### Actual result:
As the model overfits it encounters a wrong loss calculation that yields incorrect "exploding" gradients.

### Initial ideas:
The issue is almost definitely related to the diagonal epsilon value added for Cholesky decomposition in the loss calculation, as changing the value of epsilon also changes the significance of this issue (larger epsilon yields losses that are more wrong).

My initial idea was that as standard deviations approach 0, the epsilon dominates and throws the calculation off drastically. Plotting the models predictions when it encounters a wrong loss (and before updating weights), however, does not reveal any variances close to 0. (see attached pdf above).

 It might still have to do with certain correlations approaching 0 but I have not investigated further.

**Comment by jonas-scholz123** [comment]
Hi Tom,

On the first point: yes, the loss corresponds to overfitting to a single task (the task displayed below).

On the epsilon issue:
 - To reproduce, try overfitting to a single task with low `B.epsilon` (e.g. 1e-8) and then again with a higher `B.epsilon` (e.g. 1e-2) and printing the losses when they are unrealistically low. What I found is that with a higher `B.epsilon`, the normalised losses get worse (in the negative thousands), whereas a higher `B.epsilon` gives (normalised) losses in the negative 10s.
 - The way I intuitively understand it is that `B.epsilon` adds a bias/mistake to the calculation. When the unbiased "real" variances are large, `B.epsilon` is negligible and the loss is fine. When the real variances -> 0 (the model overfits), the bias dominates the real variance and the returned loss is very far off. When `B.epsilon` is smaller, this issue is smaller in magnitude than when `B.epsilon` is larger.


Here are the key parts of the code I used for the overfitting, though admittedly this might not be useful as its tied to my own application.

```
    def overfit_train(self):
        """
        Overfit to just one task to see if the model works.
        """
        epoch_losses = []

        self.optimiser = optim.Adam(self.model.model.parameters(), lr=self.opt.lr)

        self.pbar = tqdm(range(self.start_epoch, self.opt.num_epochs + 1))

        if self.start_epoch == 1:
            val_loss = self.evaluate()
            self._log(0, val_loss, val_loss)

        for epoch in self.pbar:
            batch_losses = []
            for i in range(self.opt.batches_per_epoch):
                date = "2012-01-01"
                test_date = pd.Timestamp(date)
                task = self.task_loader(test_date, context_sampling=("all", "all"))
                if i == 0:
                    self.plot_prediction(name=f"epoch_{epoch}_test", task=task)
                try:
                    batch_loss = self.train_on_batch(task)
                except:
                    self.plot_prediction(name=f"epoch_{epoch}_test_broken", task=task)
                    return

                batch_losses.append(batch_loss)

            train_loss = np.mean(batch_losses)
            epoch_losses.append(train_loss)
            val_lik = self.evaluate()
            self._log(epoch, train_loss, val_lik)
            self._save_weights(epoch, val_lik)
```

with
```
    def train_on_batch(self, tasks):
        if not isinstance(tasks, list):
            tasks = [tasks]
        self.optimiser.zero_grad()
        task_losses = []
        for task in tasks:
            task_losses.append(self.model.loss_fn(task, normalise=True))
        mean_batch_loss = B.mean(torch.stack(task_losses))
        mean_batch_loss.backward()
        torch.nn.utils.clip_grad_norm_(self.model.model.parameters(), 0.01)

        l = float(mean_batch_loss.detach().cpu().numpy())
        if l < -5:
            print("Bad loss!")
            self.plot_prediction(name=f"loss_{l}", task=task)
        self.optimiser.step()
        return l
```

**Comment by tom-andersson** [comment]
Hey @jonas-scholz123, apologies for the delay and thanks for all the extra info. I tried to overfit to a single task with a different model/data set-up, but wasn't able to reproduce your findings of spiking loss when overfitting, see below:

![epsilon_default](https://github.com/tom-andersson/deepsensor/assets/26459412/8b5d08df-dccb-4dd8-b3c8-844deff114c1)

I also tried using your two methods above (with some bits cut out) and didn't get the same findings, even with `B.epsilon = 1e-12`. 

I'll close this for now but your experiments above may be helpful if people have issues with the ConvGNP in future - so thank you!
